### ISSUE TITLE
Can't reproduce build version 3.3.0 probably due to missing tag

### ISSUE THREAD

**Issue by Giszmo** [Blockstream/green_android#75]
Unfortunately I'm in the process of reverting [my review of Green Wallet](https://walletscrutiny.com/posts/2019/11/greenwallet/) from being verifiable to not verifiable as I fail to reproduce the latest build. This might be a missing tag, only.

The new review in full:

```
$ cd green_android
$ git pull
$ git tag | grep 3.3.0
```

There is no such tag. Empty output.

```
$ git log
...
commit 820b79780ea9bc447f2794d4d34a64f5796a2d55
Author: Riccardo Casatta <riccardo@casatta.it>
Date:   Tue Dec 17 18:28:41 2019 +0100

    increment to version 3.3.0
```

We will try that one:

```
$ docker run -it --volume $PWD:/mnt --workdir /mnt --rm mycelium-wallet bash
root@efae6adea0d1:/mnt# yes | /opt/android-sdk/tools/bin/sdkmanager "build-tools;29.0.2"
root@efae6adea0d1:/mnt# git checkout 820b79780ea9bc447f2794d4d34a64f5796a2d55
root@efae6adea0d1:/mnt# ./gradlew clean build -x test
...
> Task :app:stripProduction_21DebugDebugSymbols
Unable to strip library '/mnt/app/build/intermediates/merged_native_libs/production_21Debug/out/lib/arm64-v8a/libgreenaddress.so' due to missing strip tool for ABI 'ARM64_V8A'. Packaging it as is.
Unable to strip library '/mnt/app/build/intermediates/merged_native_libs/production_21Debug/out/lib/x86_64/libgreenaddress.so' due to missing strip tool for ABI 'X86_64'. Packaging it as is.
Unable to strip library '/mnt/app/build/intermediates/merged_native_libs/production_21Debug/out/lib/x86/libgreenaddress.so' due to missing strip tool for ABI 'X86'. Packaging it as is.
Unable to strip library '/mnt/app/build/intermediates/merged_native_libs/production_21Debug/out/lib/armeabi-v7a/libgreenaddress.so' due to missing strip tool for ABI 'ARMEABI_V7A'. Packaging it as is.
...
BUILD SUCCESSFUL in 3m 37s
```

This looks like the APK would be smaller had we installed some tool to achieve
that? That sounds potentially bad for deterministic builds. Anyway:

```
root@efae6adea0d1:/mnt# exit
$ apktool d -o fromGoogle Green\ 3.3.0\ \(com.greenaddress.greenbits_android_wallet\).apk 
$ apktool d -o fromBuild app/build/outputs/apk/production/release/app-production-release-unsigned.apk 
$ diff --brief --recursive from*
Files fromBuild/apktool.yml and fromGoogle/apktool.yml differ
Files fromBuild/lib/arm64-v8a/libgreenaddress.so and fromGoogle/lib/arm64-v8a/libgreenaddress.so differ
Files fromBuild/lib/armeabi-v7a/libgreenaddress.so and fromGoogle/lib/armeabi-v7a/libgreenaddress.so differ
Files fromBuild/lib/x86/libgreenaddress.so and fromGoogle/lib/x86/libgreenaddress.so differ
Files fromBuild/lib/x86_64/libgreenaddress.so and fromGoogle/lib/x86_64/libgreenaddress.so differ
Only in fromGoogle/original/META-INF: GREENADD.RSA
Only in fromGoogle/original/META-INF: GREENADD.SF
Files fromBuild/original/META-INF/MANIFEST.MF and fromGoogle/original/META-INF/MANIFEST.MF differ
Files fromBuild/smali/com/blockstream/libgreenaddress/GDK.smali and fromGoogle/smali/com/blockstream/libgreenaddress/GDK.smali differ
```

That does not look good. There is also `production_21`, not sure what it is:

```
$ rm -rf fromBuild/
$ apktool d -o fromBuild app/build/outputs/apk/production_21/release/app-production_21-release-unsigned.apk 
$ diff --brief --recursive from* | wc -l
1202
```

1202 files differing is not good.

Unfortunately with the missing tag we fail to verify the build and have to revert
our prior verdict to **not verifiable**.

**Comment by RCasatta** [comment]
The tag has been pushed, sorry about that. We are looking to improve the process, will let you know.

**Comment by Giszmo** [comment]
Are you sure this is the right revision? It's the one I actually had guessed, so I got the same results: not verifiable.

If you are sure, could you please provide appropriate build instructions to verify the build? As written in my updated review, I had compiled the successful verification on my host system but as that is not well defined and ... my host system where I don't want to run random code, I now switched to a Docker image based on `ubuntu:18.04`.

**Comment by lvaccaro** [comment]
Hi @Giszmo, have you tried to update gdk library?

You can manually run `./app/fetch_gdk_binaries.sh` script to update gdk binaries from https://github.com/Blockstream/gdk: green release_3.3.0 uses gdk v0.0.24 and the released gdk binaries are available at [here](https://github.com/Blockstream/gdk/releases/tag/release_0.0.24), for android package `gdk-android-jni.tar.gz`. (Documentation should be improved)

The build process is described in [.gitlab-ci.yml](https://github.com/Blockstream/green_android/blob/master/.gitlab-ci.yml), command: `./gradlew clean assembleProductionRelease` should works fine.

I think `production_21` is a productFlavors to speed up build process on debug and testing, not for release, because the published apk uses minSdkVersion 19.

Please let us know if you can verify. Thank you

**Comment by Giszmo** [comment]
Now the wallet is verifiable again. Thanks for your cooperation!

I hope to soon automate things, to eliminate human error and would very much appreciate any help from projects that obviously care about verifiability, like you guys, with this. My hope is that you also run the same script and thereby make sure to not get false alerts with WalletScrutiny over nothing.

My plan is to provide an (open source and verifiable) update monitor app that checks every update of a known wallet app with the walletscrutiny server by asking it if the sha256sum(wallet.apk) is known. If not, the app would upload it for analysis. I have no better idea of how to get updates early on, especially if the provider launches the update in a different region or a different device, first.

Later the server would run the app's script on it, extracting the versionName, checking out the tag that matches the versionName, compiling it, ...

**Comment by domegabri** [comment]
Marked as enhancement to improve build process verifiability and documentation.

**Comment by Giszmo** [comment]
3.3.9 has a tiny diff which unfortunately makes the rebuild fail. It looks like the png tool is not deterministic as `app/src/main/res/drawable/ic_{testnet_btc_no_padding,sent}.xml` give a diff on my end. I updated [my results](https://walletscrutiny.com/android/com.greenaddress.greenbits_android_wallet/) accordingly.

**Comment by domegabri** [comment]
thank you for the report, we are working on making it reproducible

**Comment by domegabri** [comment]
hello @Giszmo , a potential fix for this has been merged in master and released with 3.4.0  (68b03562affd64d20dfbf86f94487ee4a506bb98.

**Comment by Giszmo** [comment]
That worked. Thanks.
